source/Variant.cpp

```diff
@@ -0,0 +1,303 @@
+/* Variant.cpp
+Copyright (c) 2019 by Michael Zahniser
+
+Endless Sky is free software: you can redistribute it and/or modify it under the
+terms of the GNU General Public License as published by the Free Software
+Foundation, either version 3 of the License, or (at your option) any later version.
+
+Endless Sky is distributed in the hope that it will be useful, but WITHOUT ANY
+WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR A
+PARTICULAR PURPOSE.  See the GNU General Public License for more details.
+*/
+
+#include "Fleet.h"
+
+#include "DataNode.h"
+#include "GameData.h"
+#include "Random.h"
+#include "Ship.h"
+
+#include <algorithm>
+#include <cmath>
+#include <iterator>
+
+using namespace std;
+
+
+
+// Construct and Load() at the same time.
+Variant::Variant(const DataNode &node)
+{
+	Load(node);
+}
+
+
+
+void Variant::Load(const DataNode &node, const bool global)
+{
+	// If this variant is global (i.e. a root node variant stored in GameData),
+	// it must be named.
+	if(global && node.Size() < 2)
+	{
+		node.PrintTrace("No name specified for variant:");
+		return;
+	}
+	if(node.Size() >= 2 && !node.IsNumber(1))
+		name = node.Token(1);
+	
+	// If Load() has already been called once on this variant, any subsequent
+	// calls will replace the contents instead of adding to them.
+	bool reset = !variants.empty() || !stockVariants.empty() || !ships.empty();
+	
+	for(const DataNode &child : node)
+	{
+		bool add = (child.Token(0) == "add");
+		bool remove = (child.Token(0) == "remove");
+		bool variant = (child.Token(add || remove) == "variant");
```

Comment:
Should this first have a check if Token(1) exists if add or remove was given?

Reply:
I don't believe it would be possible to even reach this point if Token(1) didn't exist. Variant::Load isn't going to be called by a line that just reads `add` or `remove`.